### ISSUE TITLE
WSJ downloads failing because of anti-scraping mechanism

### ISSUE THREAD

**Issue by thisisparker** [thisisparker/xword-dl#178]
WSJ is returning 401/403 errors to requests from `requests`, including `xword-dl`. My guess is that this is in response to traffic patterns they're seeing and they will turn it off again in due course, but that's a waiting game.

In the meantime, the error message should probably differentiate between this kind of connection error and a parsing error (which is what everything sounds like now).

**Comment by thisisparker** [comment]
Looking into it: this appears to be operated by a company called Datadome and they're setting and checking a cookie called `datadome` with a long token value. Theoretically we could provide that value with requests similar to an auth token, but I'd rather not have to do that. Still hoping this is temporary!

**Comment by thisisparker** [comment]
Maybe fixed this with #183, though I'm not thrilled with maintaining a list of random cookies that are required for each site and I don't know how long `datadome` cookies last anyway. Leaving open for now :roll_eyes:

**Comment by thisisparker** [comment]
Unsurprisingly, `datadome` tokens turn out to be very short-lived—on the order of hours, I guess? Maybe back to the drawing board here.

**Comment by crosswordnexus** [comment]
You don't want to just pull from Martin Herbach's site?
http://herbach.dnsalias.com/wsj/wsj240720.puz

**Comment by thisisparker** [comment]
Nope, not in `xword-dl` itself. Obviously that's a good option for end users who want it, but I've made the design decision that this tool only uses first-party sources and does its own scraping and parsing.

**Comment by lcpdecastro** [comment]
**TLDR:** `https://api.puzzlr.net/trpc/crossword.getLevel?batch=1&input=%7B%220%22%3A%7B%22tenant%22%3A%22wsj%22%2C%22date%22%3A%22<date in YYYY-MM-DD>%22%7D%7D` *(date must be from 2026-06-08 or later)*

---

Update: It seems that WSJ puzzles can now be retrieved a lot more easily than before.

Some time in the last few weeks, WSJ introduced a newer applet for their puzzles. While users on the website have been complaining about how difficult it is to navigate the crossword grid (and I can see why), I was trying to figure out how difficult it was to get the data for the crosswords. Turns out, it's a lot easier than it used to be.

With the old applet, the data for the puzzles used to be stored under the WSJ domain. It wasn't affected by anti-scraping mechanisms, but the IDs assigned to the puzzles were effectively unguessable and would have required scraping the actual puzzle's article to get, which wouldn't have worked.

Now, puzzles are hosted on a third-party API. Checking fetch calls when loading a puzzle shows that the applet fetches data from a URL that follows this pattern (percent entities have been converted for easier readability):

`https://api.puzzlr.net/trpc/crossword.getLevel?batch=1&input={"0":{"tenant":"wsj","shortId":"<insert puzzle ID here>"}}`

I thought, "darn, a puzzle ID I'd have to scrape the actual article for". But then, looking through the fetch response, it featured a JSON object with the `shortId` field and the same ID as in the URL parameter, as well as a `date` field. Considering that the API is most probably searching a database using the `shortID`, I tried using the `date` as the database search parameter instead. Surprisingly, it worked, and we can now retrieve a puzzle for a certain date using the URL below (remember to convert special characters to percent entities):

`https://api.puzzlr.net/trpc/crossword.getLevel?batch=1&input={"0":{"tenant":"wsj","date":"<insert date here in YYYY-MM-DD>"}}`

Unfortunately, the earliest puzzle in its database is from 8 June 2026, Monday (probably the first day they used the new applet). On the main crosswords landing page (https://www.wsj.com/games/crosswords), the earliest puzzle that appears when clicking "Load More" will be from that date. Older puzzles haven't been migrated to this new applet, and there's no telling if they'll ever be.

Maybe `xword-dl` can (at least to some extent) support WSJ again now?

**Comment by boisvert42** [comment]
https://github.com/thisisparker/xword-dl/pull/350

@thisisparker It's failing the tests but that's because of other outlets. WSJ tests pass fine. I suppose we need another PR to either disable some of those tests or fix those outlest.

**Comment by lcpdecastro** [comment]
Some additional things I've noticed this past week:
* Contest crosswords (Friday crosswords) are initially published with no solutions (i.e., `answer` fields in the JSON data are empty strings). However, contest crosswords from 3 July 2026 and later have been updated with answers after the contest period, which I've read ends at Sunday midnight (not 100% sure on that though). Perhaps this could be taken note of, since crossword applications might do some weird things when answers to cells are empty strings.
  * Maybe prevent downloading a puzzle when all white cells have empty solutions? In WSJ's case, the JSON data has an extra `contestClosed` field that is `true` once the contest is over and is only present in contest crosswords; this could be a good way to check instead of looping through all cells and checking if they're all black or have empty solutions.
* As of 12 July 2026 (Sunday), puzzles have been scheduled on their database one day early, causing puzzles to be antedated by one day as well (see https://www.reddit.com/r/crossword/comments/1uy0gxc/comment/oy56rlh/). Whether this gets reverted in the future is uncertain, though as of the writing of this comment (20 July 2026) this is still the case.
  * For reference, trying to access today's puzzle returns `[{"result":{"data":{"unavailable":true,"reason":"future_date","date":"2026-07-20"}}}]`. The puzzle is only fetched if I set the date to 19 July 2026.
